syntax = "proto3";

package university;

option java_package = "com.university.library";
option java_outer_classname = "UniversityProto";

// Book entity
message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  string publisher = 5;
  int32 pageCount = 6;
  int32 stock = 7;
}

// Student entity
message Student {
  string id = 1;
  string name = 2;
  string studentNumber = 3;
  string email = 4;
  bool isActive = 5;
}

// Loan entity
enum LoanStatus {
  ONGOING = 0;
  RETURNED = 1;
  LATE = 2;
}

message Loan {
  string id = 1;
  string studentId = 2;
  string bookId = 3;
  string loanDate = 4;
  string returnDate = 5;
  LoanStatus status = 6;
}

// Service definitions
service BookService {
  rpc ListBooks(Empty) returns (stream Book);
  rpc GetBook(GetByIdRequest) returns (Book);
  rpc AddBook(Book) returns (OperationStatus);
  rpc UpdateBook(Book) returns (OperationStatus);
  rpc DeleteBook(GetByIdRequest) returns (OperationStatus);
}

service StudentService {
  rpc ListStudents(Empty) returns (stream Student);
  rpc GetStudent(GetByIdRequest) returns (Student);
  rpc AddStudent(Student) returns (OperationStatus);
  rpc UpdateStudent(Student) returns (OperationStatus);
  rpc DeleteStudent(GetByIdRequest) returns (OperationStatus);
}

service LoanService {
  rpc ListLoans(Empty) returns (stream Loan);
  rpc GetLoan(GetByIdRequest) returns (Loan);
  rpc BorrowBook(Loan) returns (OperationStatus);
  rpc ReturnBook(GetByIdRequest) returns (OperationStatus);
}

// Common messages
message GetByIdRequest {
  string id = 1;
}

message OperationStatus {
  bool success = 1;
  string message = 2;
}

message Empty {}
